core/src/main/java/org/apache/iceberg/encryption/AesGcmInputStream.java

```diff
@@ -0,0 +1,212 @@
+  private int lastBlockSize;
+  private long plainStreamSize;
+  private byte[] fileAadPrefix;
+
+  AesGcmInputStream(SeekableInputStream sourceStream, long sourceLength,
+                    byte[] aesKey, byte[] fileAadPrefix) throws IOException {
+    this.netSourceFileSize = sourceLength - AesGcmOutputStream.PREFIX_LENGTH;
+    this.sourceStream = sourceStream;
+    byte[] prefixBytes = new byte[AesGcmOutputStream.PREFIX_LENGTH];
+    int fetched = sourceStream.read(prefixBytes);
+    Preconditions.checkArgument(fetched == AesGcmOutputStream.PREFIX_LENGTH,
+        "Insufficient read " + fetched);
+    this.plainStreamPosition = 0;
+    this.fileAadPrefix = fileAadPrefix;
+
+    byte[] magic = new byte[AesGcmOutputStream.MAGIC_ARRAY.length];
+    System.arraycopy(prefixBytes, 0, magic, 0, AesGcmOutputStream.MAGIC_ARRAY.length);
+
+    Preconditions.checkArgument(Arrays.equals(AesGcmOutputStream.MAGIC_ARRAY, magic),
+        "File with wrong magic string. Should start with " + AesGcmOutputStream.MAGIC_STRING);
+
+    plainBlockSize = ByteBuffer.wrap(prefixBytes, AesGcmOutputStream.MAGIC_ARRAY.length, 4)
+        .order(ByteOrder.LITTLE_ENDIAN).getInt();
+    cipherBlockSize = plainBlockSize + AesGcmOutputStream.GCM_NONCE_LENGTH + AesGcmOutputStream.GCM_TAG_LENGTH;
+
+    try {
+      gcmCipher = Cipher.getInstance("AES/GCM/NoPadding");
+    } catch (GeneralSecurityException e) {
+      throw new IOException(e);
+    }
+    this.nonce = new byte[AesGcmOutputStream.GCM_NONCE_LENGTH];
+    this.key = new SecretKeySpec(aesKey, "AES");
+    this.ciphertextBlockBuffer = new byte[cipherBlockSize];
+    this.currentBlockIndex = 0;
+    this.currentOffsetInPlainBlock = 0;
+
+    numberOfBlocks = (int) (netSourceFileSize / cipherBlockSize);
+    lastBlockSize = (int) (netSourceFileSize % cipherBlockSize);
+    if (lastBlockSize == 0) {
+      lastBlockSize = cipherBlockSize;
+    } else {
+      numberOfBlocks += 1;
+    }
+
+    plainStreamSize = (numberOfBlocks - 1L) * plainBlockSize +
+            (lastBlockSize - AesGcmOutputStream.GCM_NONCE_LENGTH - AesGcmOutputStream.GCM_TAG_LENGTH);
+  }
+
+  public long plaintextStreamSize() {
+    return plainStreamSize;
+  }
+
+  @Override
+  public int available() throws IOException {
+    return Math.toIntExact(plainStreamSize - plainStreamPosition);
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    if (len <= 0) {
+      throw new IOException("Negative read length " + len);
+    }
+
+    if (available() <= 0) {
+      return -1;
+    }
+
+    boolean lastBlock = currentBlockIndex + 1 == numberOfBlocks;
+    int resultBufferOffset = off;
+    int remaining = len;
+
+    sourceStream.seek(AesGcmOutputStream.PREFIX_LENGTH + currentBlockIndex * cipherBlockSize);
+
+    while (remaining > 0) {
+      int toLoad = lastBlock ? lastBlockSize : cipherBlockSize;
```

Comment:
This is part of the code I think is kind of complicated because of the special casing of the last block logic. Let's try to make this as clear as possible. 

Do we actually need to keep these numbers here? I was thinking about this again, could we instead do something like

Math.min(cipherBlockSize, remaining)?